### PR TITLE
 Feedback session results download should support richer format than csv #5801 

### LINKED ISSUE #5801: Feedback session results download should support richer format than csv
The idea of this issue is to explore whether we can support richer format such as xls (excel) in feedback session results download. Alternatively, we should look in to the option of opening the file as GDoc.

### PR DESCRIPTION
**Pull request by rclakmal**
Fixes #5801

This PR is used to discuss addition of richer download formats to feedback results.  This is how the two options look like. Suggestions on how to phrase these options better will be helpful. 

snip20160712_18

### CODE REVIEW

**Comment by damithc** [comment]
May be use this phrasing?
```
As normal spreadsheet (.xls file)
As spreadsheet without formatting (.csv file)
```

**Comment by rclakmal** [comment]
Changed wordings. This is open for review now.

File: src/main/java/teammates/ui/controller/FileDownloadResult.java [hunk 1]
```diff
@@ -69,5 +51,9 @@ public String getFileName() {
     public String getFileContent() {
         return this.fileContent;
     }
+    
+    public String getFileType(){
```
Comment (chowyb) [inline src/main/java/teammates/ui/controller/FileDownloadResult.java]: No whitespace between `()` and `{` here, causing checkstyle violations. Should be possible to review this once it is fixed and the build passes.

**Comment by rclakmal** [comment]
Fixed.

**Comment by kanghj** [comment]
@rclakmal, I think this is strange. Although the file extension is .xls, the data is still csv. So this doesn't enable us to add text formatting, right? Will you be extending this in future PRs?

**Comment by rclakmal** [comment]
@kanghj Currently the content of the file is the same as the one we generate for csv. But since contentType have been changed, in the future we could add more formatting (or extra items such as charts) as needed to XSL version.  @damithc I believe this was the idea ?

This PR only enable downloads of multiple contentTypes (`text/csv` and `application/vnd.ms-excel`). This way we could even extend to more download types if needed in the future.

**Comment by damithc** [comment]
Yes that's the idea. Perhaps add a small formatting change (e.g. make a column wider) in this PR itself to show that formatting changes are possible?

**Comment by rclakmal** [comment]
@damithc @kanghj For formatting purposes we should look for a suitable library such as [Apache POI](https://poi.apache.org/) or [JExcel](https://www.teamdev.com/jexcel) (which is a commercial product but provides free license for open source or academic projects). We can easily integrate the option we choose in the newly introduced `XLSFileDownloadResult` class.

However, evaluating these options would be nice to take up in another PR.

**Comment by damithc** [comment]
Umm... does that mean the current xls file in fact a csv file with an incorrect extension?

**Comment by kanghj** [comment]
@rclakmal, in fact, I can't even open the .xls file after downloading it. We should try to add the suitable library either in this PR or a different PR before we can merge this.

**Comment by rclakmal** [comment]
@damithc Nope, it has the correct contentType  for an excel file. (so internal contentType and both extension are different)

@kanghj That is strange. I can open it fine with 2016 Excel here. Do you get an error message ?

However, I will do also the integration of the library in this PR then.

**Comment by kanghj** [comment]
Here is the error I get:

excelerror

**Comment by rclakmal** [comment]
@kanghj @damithc Introduced [Apache POI ](https://poi.apache.org/). 

Apache POI wins across the features, documentation and community support against the [JExcel](http://jexcelapi.sourceforge.net/). And JExcel lacks support for recent versions of excel. 

Introduced a test formatting (Questions are bold and bigger) as a proof-of-concept.

File: src/main/java/teammates/ui/controller/CSVFileDownloadResult.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import java.io.PrintWriter;
+import java.util.List;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import teammates.common.datatransfer.AccountAttributes;
+import teammates.common.util.StatusMessage;
+
+public class CSVFileDownloadResult extends FileDownloadResult {
```
Comment (damithc) [inline src/main/java/teammates/ui/controller/CSVFileDownloadResult.java]: Should be `CsvFileDownloadResult`

File: src/main/java/teammates/ui/controller/XLSFileDownloadResult.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        outStream.close();
+    }
+
+    private XSSFWorkbook getWorkBook() {
+        workBook = new XSSFWorkbook();
+        XSSFSheet sheet = workBook.createSheet("sheet1");
+        int rowNumber = 0;
+        String[] lines = fileContent.split(Const.EOL);
+        for (String line : lines) {
+            String[] str = line.split(",");
```
Comment (kanghj) [inline src/main/java/teammates/ui/controller/XLSFileDownloadResult.java]: What if the text contains a `,` in it, won't this incorrect split it up?

**Comment by rclakmal** [comment]
@damithc Fixed file naming.

@kanghj Thanks for the catch. Updated the regex to fix that.

File: src/main/java/teammates/ui/controller/XlsFileDownloadResult.java [hunk 4]
```diff
@@ -44,7 +44,8 @@ private XSSFWorkbook getWorkBook() {
         int rowNumber = 0;
         String[] lines = fileContent.split(Const.EOL);
         for (String line : lines) {
-            String[] str = line.split(",");
+            // Split by comma only if comma has a zero or even number of quotes in front of it.
+            String[] str = line.split(",(?=([^\"]*\"[^\"]*\")*[^\"]*$)", -1);
```
Comment (damithc) [inline src/main/java/teammates/ui/controller/XlsFileDownloadResult.java]: We should unit test regexes as they are sources of subtle errors.  Define it as a constant and then use the same constant in a unit test. I think there are examples of regex unit tests somewhere.

File: src/main/java/teammates/ui/controller/XlsFileDownloadResult.java [hunk 5]
```diff
@@ -19,10 +19,10 @@
 import teammates.common.util.Const;
 import teammates.common.util.StatusMessage;
 
-public class XLSFileDownloadResult extends FileDownloadResult {
+public class XlsFileDownloadResult extends FileDownloadResult {
     private XSSFWorkbook workBook;
 
-    public XLSFileDownloadResult(String destination, AccountAttributes account, List<StatusMessage> status,
+    public XlsFileDownloadResult(String destination, AccountAttributes account, List<StatusMessage> status,
```
Comment (damithc) [inline src/main/java/teammates/ui/controller/XlsFileDownloadResult.java]: There are other places in this file where a similar fix should be applied. e.g. XSSFWorkbook

Comment (rclakmal) [inline src/main/java/teammates/ui/controller/XlsFileDownloadResult.java]: @damithc Can't help with those. Coming from the library itself :)

Comment (damithc) [inline src/main/java/teammates/ui/controller/XlsFileDownloadResult.java]: Scratch that. Just realized it's a third-party class.

File: src/main/java/teammates/ui/controller/XlsFileDownloadResult.java [hunk 6]
```diff
@@ -44,7 +44,8 @@ private XSSFWorkbook getWorkBook() {
         int rowNumber = 0;
```
Comment (damithc) [inline src/main/java/teammates/ui/controller/XlsFileDownloadResult.java]: As much as possible, try to write standalone methods. i.e. all input come as parameters and the only output is the return value. Doing so improves testability. For example, you can have a utility method `convertToWorkbook` that takes a string as the input and returns the corresponding workbook. That method should be unit tested to ensure it works for all cases. empty content, comma inside cell, etc.

**Comment by rclakmal** [comment]
In-cooperated review comments. Added more unit testing.

File: src/main/java/teammates/ui/controller/XlsxFileDownloadResult.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+    public XSSFWorkbook createWorkBook() {
+        workBook = new XSSFWorkbook();
+        XSSFSheet sheet = workBook.createSheet("sheet1");
+        int rowNumber = 0;
+        if (!Strings.isNullOrEmpty(fileContent)) {
+            String[] lines = fileContent.split(Const.EOL);
+            for (String line : lines) {
+                // Split by comma only if comma has a zero or even number of quotes infront.
+                String[] str = line.split(CSV_LINE_DELIMITTER_REGEX, -1);
```
Comment (kanghj) [inline src/main/java/teammates/ui/controller/XlsxFileDownloadResult.java]: Hmm, this still seems a little fragile. If the question or response text contains quotations and commas, then it is likely that this approach will fail. I suspect we have to make changes at the code before `FileDownloadResult` is used. Instead of constructing csv formatted file contents then trying to convert the csv formatted contents into xls, we should try to construct the data in a different way, or construct the `XSSFWorkbook` in the `*Logic` classes.

Comment (damithc) [inline src/main/java/teammates/ui/controller/XlsxFileDownloadResult.java]: I think so too. It's more logical to build up the data in a generic format first and have different 'renderers' to render that data in different formats. csv should not be used as that generic format.

Comment (rclakmal) [inline src/main/java/teammates/ui/controller/XlsxFileDownloadResult.java]: Well, all variable contents are sanitized before coming to the state (to handle commas in csv file). And with the updated regex, we are ignoring commas which have been sanitized. So I don't think approach will fail. We just have to  test for more corner cases

However, I agree, we could better structure the code. I already started having a look and restructuring.  But it turns out to be very big change than expected. Almost all methods for file download are written in CSV specific way. Not in a generic way. So this means we have port all these methods to xls. Few examples. 

`getCsvDetailedResponsesRow`, `getRemainingRowsInCsvFormat`, and  'getQuestionResultStatisticsCsv'  which is overridden in all question types.  These are few examples.

**Comment by damithc** [comment]
The problem with csv is it is a minimal format. That means it cannot contain all the details. A good example of this problem we are going to encounter soon is rich text. The csv format will not contain rich text formatting and any other formats generated from csv cannot support rich text as well.

**Comment by rclakmal** [comment]
@damithc True, csv cannot support rich formats. But my point was for text content we can keep csv as the generic format and then on top of that add formatting rules depending on the download type.

**Comment by damithc** [comment]
The problem I'm anticipating is when we add rich text support to responses (due to happen soon). When you generate the csv, all html tags will be stripped out form the responses. The plain text csv may be enough to generate xls file but may be not enough to generate a richer format such as html or pdf. But as we have only csv and xls formats, I'm ok to go with the current solution for now. @kanghj what do you think?

**Comment by kanghj** [comment]
Ok, let's go with the current solution now then. I guess we will eventually need to move most of the code to generate an output more generic than csv.

File: build.gradle [hunk 8]
```diff
@@ -49,7 +49,9 @@ dependencies {
                     "com.sun.jersey:jersey-core:1.19",
                     "com.sun.jersey.contribs:jersey-multipart:1.19",
                     "org.apache.taglibs:taglibs-standard-impl:1.2.5",
-                    "org.jsoup:jsoup:1.9.2"
+                    "org.jsoup:jsoup:1.9.2",
+                    "org.apache.poi:poi:3.9",
```
Comment (kanghj) [inline build.gradle]: Is there any particular reason why these versions are chosen. Latest stable version appears to be 3.14?

Comment (rclakmal) [inline build.gradle]: The reason was the [usage](https://mvnrepository.com/artifact/org.apache.poi/poi) of versions. This version had the highest current usage up to now. Shall I change to the latest ?

**Comment by rclakmal** [comment]
@damithc @kanghj I see your point now. I'm open to make the refactoring. But can someone more familiar with the legacy code, go through for a while and suggest if there is a more suitable way to overcome the challenges i mentioned above. 

@kanghj are we already striping the html tags when preparing the csv data ? Could you show me where it is done ?

**Comment by kanghj** [comment]
@rclakmal, it is not currently done yet, but that should happen soon. Adding rich text support to the different parts of the project is still ongoing.

**Comment by kanghj** [comment]
@rclakmal, do you intend to try to make that refactoring in this PR?

**Comment by rclakmal** [comment]
I suggest we do it once we have rich text functionality implemented. Because then we can figure out the most suitable way of handing this problem @damithc  brought out. But I'm happy to hear your view too.

**Comment by kanghj** [comment]
Actually, the question description and session instructions can already support rich text, although they are not included in the downloaded report. If you want to experiment around first, you can try including the session instructions in the excel file.

File: build.gradle (hunk 8 above)
Comment (kanghj) [inline build.gradle]: I think it should be better if we use the latest version since it's stable already, unless we find any reason why 3.9 is better.

Comment (rclakmal) [inline build.gradle]: Upgraded

**Comment by rclakmal** [comment]
@kanghj Would like to take up that. But may be through another issue/PR ?

**Comment by kanghj** [comment]
@rclakmal, sure, go ahead and make a new issue for that then

**Comment by rclakmal** [comment]
@kanghj May I know why this PR was put onhold ?

**Comment by kanghj** [comment]
Sorry, I thought you were wanted to wait for rich text to be implemented for responses first

**Comment by rclakmal** [comment]
I believe it has no dependency on each other. But yes, that's pending on my list. 
However, @damithc has given it a [low](#5898) priority in the tracker. Hence I'm currently looking in to #5859.

File: src/main/java/teammates/ui/controller/XlsxFileDownloadResult.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+        workBook.write(outStream);
+        outStream.flush();
+        outStream.close();
+    }
+
+    public XSSFWorkbook createWorkBook() {
+        workBook = new XSSFWorkbook();
+        XSSFSheet sheet = workBook.createSheet("sheet1");
+        int rowNumber = 0;
+        if (!Strings.isNullOrEmpty(fileContent)) {
```
Comment (kanghj) [inline src/main/java/teammates/ui/controller/XlsxFileDownloadResult.java]: We should be able to omit this.

File: src/test/java/teammates/test/cases/ui/InstructorFeedbackResultsDownloadActionTest.java [hunk 10]
```diff
@@ -76,6 +83,21 @@ public void testExecuteAndPostProcess() throws Exception {
...
+        
+        expectedDestination = "filedownload?" + "error=false" + "&user=idOfInstructor1OfCourse1";
+        assertEquals(expectedDestination, result.getDestinationWithParams());
+        assertFalse(result.isError);
+        assertEquals("", result.getStatusMessage());
+
+        expectedFileName = session.getCourseId() + "_" + session.getFeedbackSessionName();
+        assertEquals(expectedFileName, result.getFileName());
+        assertEquals(Const.FeedbackSessionResultsDownloadTypes.XLSX, result.getFileType());
+        verifyFileContentForSession1InCourse1(result.getFileContent(), session);
```
Comment (kanghj) [inline src/test/java/teammates/test/cases/ui/InstructorFeedbackResultsDownloadActionTest.java]: Are we able to test anywhere that the formatting is correct?

File: src/main/java/teammates/ui/controller/XlsxFileDownloadResult.java (hunk 9 above)
Comment (rclakmal) [inline src/main/java/teammates/ui/controller/XlsxFileDownloadResult.java]: Can you explain why ?

Comment (kanghj) [inline src/main/java/teammates/ui/controller/XlsxFileDownloadResult.java]: Is there any reason why it should be empty or null? It looks like a programmer error to me if it's empty or null so we should fail fast instead of silencing the error. It should instead be an assertion using `Assumption.*` if there isn't any valid case where it should be empty

Comment (rclakmal) [inline src/main/java/teammates/ui/controller/XlsxFileDownloadResult.java]: Included an `Assumption` to check `fileContent` null status. Removed `if` condition.

File: src/test/java/teammates/test/cases/ui/XlsxFileDownloadResultTest.java [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    @Test
+    public void testQuestionFormattingChanges() throws IOException {
+        String fileContent = "one,two" + Const.EOL + "Question 1,"
+                + Sanitizer.sanitizeForCsv("Hi What do you think about Calculus, Differential Classes ?");
+        XlsxFileDownloadResult download = new XlsxFileDownloadResult(
+                "/page/instructorFeedbackResultsDownload",
+                null, null, "Normal name",
+                fileContent, Const.FeedbackSessionResultsDownloadTypes.XLSX);
+        assertEquals(2, download.getWorkBook().getSheetAt(0).getLastRowNum());
+        assertEquals(2, download.getWorkBook().getSheetAt(0).getRow(1).getLastCellNum());
```
Comment (kanghj) [inline src/test/java/teammates/test/cases/ui/XlsxFileDownloadResultTest.java]: Let's extract the repeated `download.getWorkBook().getSheetAt(0)` to a local variable to improve readability.

**Comment by kanghj** [comment]
After this is merged, remember to make a new issue for refactoring the related code to be more generic and expressive

File: src/test/java/teammates/test/cases/ui/XlsxFileDownloadResultTest.java [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+        XlsxFileDownloadResult download = new XlsxFileDownloadResult(
+                "/page/instructorFeedbackResultsDownload",
+                null, null, "Normal name",
+                fileContent, Const.FeedbackSessionResultsDownloadTypes.XLSX);
+        assertEquals(3, download.getWorkBook().getSheetAt(0).getLastRowNum());
+        assertEquals(2, download.getWorkBook().getSheetAt(0).getRow(1).getLastCellNum());
+        assertEquals(2, download.getWorkBook().getSheetAt(0).getRow(2).getLastCellNum());
+        // Third row should only have one cell since it a single word with a
+        // comma included.
+        assertEquals(1, download.getWorkBook().getSheetAt(0).getRow(3).getLastCellNum());
```
Comment (kanghj) [inline src/test/java/teammates/test/cases/ui/XlsxFileDownloadResultTest.java]: Are we able to test the contents of the cell?

File: src/test/java/teammates/test/cases/ui/XlsxFileDownloadResultTest.java [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+                + Sanitizer.sanitizeForCsv("Hi What do you think about Calculus, Differential Classes ?");
+        XlsxFileDownloadResult download = new XlsxFileDownloadResult(
+                "/page/instructorFeedbackResultsDownload",
+                null, null, "Normal name",
+                fileContent, Const.FeedbackSessionResultsDownloadTypes.XLSX);
+        assertEquals(2, download.getWorkBook().getSheetAt(0).getLastRowNum());
+        assertEquals(2, download.getWorkBook().getSheetAt(0).getRow(1).getLastCellNum());
+        assertEquals(2, download.getWorkBook().getSheetAt(0).getRow(2).getLastCellNum());
+        assertTrue(download.getWorkBook().getSheetAt(0).getRow(2).getRowStyle().getFont().getBold());
+        assertEquals(15, download.getWorkBook().getSheetAt(0).getRow(2).getRowStyle().getFont().getFontHeightInPoints());
```
Comment (kanghj) [inline src/test/java/teammates/test/cases/ui/XlsxFileDownloadResultTest.java]: Extract a local variable for `download.getWorkBook().getSheetAt(0).getRow(2)`

**Comment by kanghj** [comment]
Some comments added. I should have no more comments after this.

**Comment by kanghj** [comment]
Unassigned myself as I will be a bit busy soon, @chowyb can take over the review

**Comment by rclakmal** [comment]
@chowyb Would you be able to review this further ?

**Comment by chowyb** [comment]
Running into the following server-side error when attempting to download an .xlsx file:

```
<b>Actual user</b>
insta@example.com
<br>

<b>Request Method</b>
POST
<br>

<b>User Agent</b>
Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.106 Safari/537.36
<br>

<b>Request Url</b>
http://localhost:8888/page/instructorFeedbackResultsDownload
<br>

<b>Request Path</b>
/page
<br>

<b>Request Parameters</b>
{frshowstats::true, frdownloadtype::xlsx, frindicatemissingresponses::true, sectionname::All, filtertext::, courseid::insta.exa-demo, user::insta@example.com, fsname::First team feedback session}
<br>

<b>Stack Trace</b>
<pre><code>

java.lang.NullPointerException
	at org.apache.poi.POIXMLDocument.write(POIXMLDocument.java:217)
	at teammates.ui.controller.XlsxFileDownloadResult.send(XlsxFileDownloadResult.java:44)
	at teammates.ui.controller.ControllerServlet.doPost(ControllerServlet.java:65)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:637)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:717)
	at org.mortbay.jetty.servlet.ServletHolder.handle(ServletHolder.java:511)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1166)
	at teammates.storage.datastore.DatastoreFilter.doFilter(DatastoreFilter.java:29)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.appengine.tools.appstats.AppstatsFilter.doFilter(AppstatsFilter.java:143)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.appengine.api.socket.dev.DevSocketFilter.doFilter(DevSocketFilter.java:74)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.appengine.tools.development.ResponseRewriterFilter.doFilter(ResponseRewriterFilter.java:128)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.appengine.tools.development.HeaderVerificationFilter.doFilter(HeaderVerificationFilter.java:34)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.appengine.api.blobstore.dev.ServeBlobFilter.doFilter(ServeBlobFilter.java:63)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.TransactionCleanupFilter.doFilter(TransactionCleanupFilter.java:43)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.appengine.tools.development.StaticFileFilter.doFilter(StaticFileFilter.java:125)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.appengine.tools.development.DevAppServerModulesFilter.doDirectRequest(DevAppServerModulesFilter.java:366)
	at com.google.appengine.tools.development.DevAppServerModulesFilter.doDirectModuleRequest(DevAppServerModulesFilter.java:349)
	at com.google.appengine.tools.development.DevAppServerModulesFilter.doFilter(DevAppServerModulesFilter.java:116)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at org.mortbay.jetty.servlet.ServletHandler.handle(ServletHandler.java:388)
	at org.mortbay.jetty.security.SecurityHandler.handle(SecurityHandler.java:216)
	at org.mortbay.jetty.servlet.SessionHandler.handle(SessionHandler.java:182)
	at org.mortbay.jetty.handler.ContextHandler.handle(ContextHandler.java:765)
	at org.mortbay.jetty.webapp.WebAppContext.handle(WebAppContext.java:418)
	at com.google.appengine.tools.development.DevAppEngineWebAppContext.handle(DevAppEngineWebAppContext.java:98)
	at org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
	at com.google.appengine.tools.development.JettyContainerService$ApiProxyHandler.handle(JettyContainerService.java:509)
	at org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
	at org.mortbay.jetty.Server.handle(Server.java:326)
	at org.mortbay.jetty.HttpConnection.handleRequest(HttpConnection.java:542)
	at org.mortbay.jetty.HttpConnection$RequestHandler.content(HttpConnection.java:938)
	at org.mortbay.jetty.HttpParser.parseNext(HttpParser.java:755)
	at org.mortbay.jetty.HttpParser.parseAvailable(HttpParser.java:218)
	at org.mortbay.jetty.HttpConnection.handle(HttpConnection.java:404)
	at org.mortbay.io.nio.SelectChannelEndPoint.run(SelectChannelEndPoint.java:409)
	at org.mortbay.thread.QueuedThreadPool$PoolThread.run(QueuedThreadPool.java:582)

</code></pre></span>|||/page/instructorFeedbackResultsDownload|||insta@example.com%20160803114857597
```

Also, the displaying of options may be a bit off:

display
The options should ideally be displayed directly under the "Download Results" button. In addition, as the options are shown as an `<a>` tag, but without a `href` attribute, the browser considers it to be plain text, and changes the cursor to the text cursor (something like a capital I), making the connection that they are actually buttons less obvious.

**Comment by rclakmal** [comment]
@chowyb Open for review. Made the list options appear close to download button and with hand cursor.

**Comment by chowyb** [comment]
@rclakmal I've identified a few issues with the current implementation, and have listed them by priority. The following issues were identified from the following screenshot I took after testing the feature.

excel

1. The use of a newline seems to generate incorrect output, as it places all text before the newline into a single cell, and text after the newline in a separate cell.
Priority: High

2. Using double quotes `"` within a value seems to add an additional double quote as well.
Priority: Medium

3. Additional double quotes seem to be added to the start and end of most inputs.
Priority: Low

**Comment by rclakmal** [comment]
@chowyb Thanks for your review comments. Could you please have a re-look for these 3 issues.

**Comment by rclakmal** [comment]
@chowyb a kind reminder on this :)

File: src/test/java/teammates/test/cases/ui/XlsxFileDownloadResultTest.java [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+
+import teammates.common.util.Const;
+import teammates.common.util.Sanitizer;
+import teammates.test.cases.BaseTestCase;
+import teammates.ui.controller.XlsxFileDownloadResult;
+
+public class XlsxFileDownloadResultTest extends BaseTestCase {
+
+    @Test
+    public void testEmptyContentXslFileDownloadResult() {
```
Comment (chowyb) [inline src/test/java/teammates/test/cases/ui/XlsxFileDownloadResultTest.java]: Might be a typo here, I suspect the intention is to use "Xls" and not "Xsl".

Applies to other methods in this class as well.

File: src/main/webapp/WEB-INF/tags/instructor/results/sessionPanel.tag [hunk 15]
```diff
@@ -45,17 +45,23 @@
...
                 <div class="form-group">
                     <div class="col-md-12">
-                        <form method="post" action="<%=Const.ActionURIs.INSTRUCTOR_FEEDBACK_RESULTS_DOWNLOAD%>">
+                        <form id="download-form" method="post" action="<%=Const.ActionURIs.INSTRUCTOR_FEEDBACK_RESULTS_DOWNLOAD%>">
                             <div id="feedbackDataButtons">
-                                <input id="button_download" type="submit" class="btn btn-primary btn-block"
-                                       name="<%=Const.ParamsNames.FEEDBACK_RESULTS_UPLOADDOWNLOADBUTTON%>"
-                                       value="Download Results">
+                                <button id="button_download" type="submit" data-toggle="dropdown" class="btn btn-primary btn-block dropdown-toggle"
+                                       name="<%=Const.ParamsNames.FEEDBACK_RESULTS_UPLOADDOWNLOADBUTTON%>">
```
Comment (chowyb) [inline src/main/webapp/WEB-INF/tags/instructor/results/sessionPanel.tag]: Align `name` with `id` here

File: src/main/webapp/WEB-INF/tags/instructor/results/sessionPanel.tag [hunk 16]
```diff
@@ -45,17 +45,23 @@
...
+                                 Download Results <span class="caret"></span>
```
Comment (chowyb) [inline src/main/webapp/WEB-INF/tags/instructor/results/sessionPanel.tag]: Indent this 4 spaces off of `<button` 2 lines above

File: src/test/java/teammates/test/cases/ui/ActionResultTest.java [hunk 17]
```diff
@@ -20,16 +21,18 @@ public void testAppendParameters() {
     
     @Test
     public void testFileDownloadResult() {
-        FileDownloadResult download = new FileDownloadResult("/page/instructorFeedbackResultsDownload",
-                                                            null, null, "Normal name", "abcabc");
+        FileDownloadResult download = new CsvFileDownloadResult("/page/instructorFeedbackResultsDownload",
+                                                            null, null, "Normal name",
+                                                            "abcabc", Const.FeedbackSessionResultsDownloadTypes.CSV);
```
Comment (chowyb) [inline src/test/java/teammates/test/cases/ui/ActionResultTest.java]: Align these two lines with `"/page/...`

File: src/test/java/teammates/test/cases/ui/ActionResultTest.java [hunk 18]
```diff
@@ -20,16 +21,18 @@ public void testAppendParameters() {
...
                      download.getContentDispositionHeader());
     }
     
     @Test
     public void testFileDownloadResult_fileNameWithUnicodeCharacters_hasCorrectContentDesposition() {
-        FileDownloadResult download = new FileDownloadResult("/page/instructorFeedbackResultsDownload",
-                                                            null, null, "£ rates", "abcabc");
+        FileDownloadResult download = new CsvFileDownloadResult("/page/instructorFeedbackResultsDownload",
+                                                            null, null, "£ rates",
+                                                            "abcabc", Const.FeedbackSessionResultsDownloadTypes.CSV);
```
Comment (chowyb) [inline src/test/java/teammates/test/cases/ui/ActionResultTest.java]: Same here (alignment)

**Comment by chowyb** [comment]
Behaviour looks good, just a few coding style changes requested.

**Comment by rclakmal** [comment]
Thanks @chowyb. Fixed styling issues.

**Comment by rclakmal** [comment]
@chowyb :)
